File: 2025/day09.py
from itertools import combinations
from shapely.geometry import Point, Polygon, LineString
from shapely import prepare

pairs = []
points = []
best = 0

with open("2025/input/day09") as f:
    data = f.read().splitlines()

    for line in data:
        x = tuple(int(x) for x in line.split(","))
        points.append(x)

for p1, p2 in combinations(points, 2):
    surface = (abs(p1[0] - p2[0]) + 1) * (abs(p1[1] - p2[1]) + 1)
    if surface > best:
        best = surface


print(f"Partie 1 : {best}")

area = Polygon([Point(x) for x in points])
prepare(area)
best = 0

for p1, p2 in combinations(points, 2):
    surface = (abs(p1[0] - p2[0]) + 1) * (abs(p1[1] - p2[1]) + 1)
    if surface <= best:
        continue
    rect_points = [
        Point(p1),
        Point(p1[0], p2[1]),
        Point(p2),
        Point(p2[0], p1[1]),
    ]
    rect = LineString(rect_points)
    if area.contains(rect):
        best = surface

print(f"Partie 2 : {best}")

File: 2025/rust/src/aoc2025/day09.rs
use geo::{Contains, LineString, Point, Polygon};
use rdcl_aoc_helpers::input::WithReadLines;
use std::fs::File;

pub fn day09(path: &String) {
    let input: Vec<String> = File::open(path).read_lines::<String>(1).collect();
    let points: Vec<Point> = input
        .iter()
        .map(|x| {
            let coords: Vec<&str> = x.split(",").collect();
            Point::new(
                coords[0].parse::<f64>().unwrap(),
                coords[1].parse::<f64>().unwrap(),
            )
        })
        .collect();

    let mut best = 0;

    for i in 0..points.len() - 1 {
        for j in i + 1..points.len() {
            let distance: u64 = (((points[i].x() - points[j].x()).abs() + 1.0)
                * ((points[i].y() - points[j].y()).abs() + 1.0))
                as u64;
            if distance > best {
                best = distance;
            }
        }
    }

    println!("Partie 1: {}", best);

    let area: Polygon = Polygon::new(LineString::from(points.clone()), vec![]);

    best = 0;

    for i in 0..points.len() - 1 {
        for j in i + 1..points.len() {
            let distance: u64 = (((points[i].x() - points[j].x()).abs() + 1.0)
                * ((points[i].y() - points[j].y()).abs() + 1.0))
                as u64;
            if distance > best {
                let p1: Point = points[i];
                let p2: Point = Point::new(points[i].x(), points[j].y());
                let p3: Point = points[j];
                let p4: Point = Point::new(points[j].x(), points[i].y());
                let edges: LineString = LineString::from(vec![p1, p2, p3, p4, p1]);
                if area.contains(&edges) {
                    best = distance;
                }
            }
        }
    }

    println!("Partie 2: {}", best);
}
